### PR TITLE
Traces jaeger

### PR DESCRIPTION
**Pull request by derek-ho**
### Description
[Describe what this change achieves]

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/event_analytics/explorer/events_views/docViewRow.tsx [hunk 1]
```diff
@@ -77,13 +77,13 @@ export const DocViewRow = forwardRef((props: IDocViewRowProps, ref) => {
...
           <dl className="source truncate-by-height">
             {toPairs(doc).map((entry: string[]) => {
-              const isTraceField = entry[0] === OTEL_TRACE_ID;
+              const isTraceField = (entry[0] === OTEL_TRACE_ID || entry[0] === JAEGER_TRACE_ID);
               return (
                 <span key={uniqueId('grid-desc')}>
                   <dt>{entry[0]}:</dt>
                   <dd>
                     <span>
-                      {isTraceField && isValidTraceId(entry[1]) && !isFlyout ? (
```
Comment (derek-ho) [inline public/components/event_analytics/explorer/events_views/docViewRow.tsx:86]: to enable event analytics traceID clicking/flyout pops out for jaeger indices

File: public/components/event_analytics/explorer/events_views/trace_block/trace_block.tsx [hunk 2]
```diff
@@ -43,5 +43,5 @@ export const TraceBlock = ({ http, hit, logTraceId }: props) => {
     );
   }
 
-  return <TraceDetailRender traceId={logTraceId} http={http} mode='data_prepper'/>;
+  return <TraceDetailRender traceId={hit.traceID || logTraceId} http={http} mode={hit.traceID === '' ? 'data_prepper' : 'jaeger'}/>;
```
Comment (joshuali925) [inline public/components/event_analytics/explorer/events_views/trace_block/trace_block.tsx]: np. would be more clear if you check it at the beginning of the function. reader might not know what's the difference between hit.traceID and logTraceId

```js
const mode = hit.traceID === '' ? 'data_prepper' : 'jaeger';
```

Comment (derek-ho) [reply public/components/event_analytics/explorer/events_views/trace_block/trace_block.tsx]: Fixed in [f3a871f](https://github.com/opensearch-project/dashboards-observability/pull/150/commits/f3a871f63d3642fa0c25a6de59c1ec6d63e3e3ef)

File: public/components/trace_analytics/requests/dashboard_request_handler.ts [hunk 3]
```diff
@@ -189,7 +190,12 @@ export const handleJaegerDashboardRequest = async (
       });
       return map;
     })
-    .catch((error) => console.error(error));
+    .catch((error) => {
+      console.log("error here")
+      console.error(error)
+      
+      setToast('hello')
```
Comment (derek-ho) [inline public/components/trace_analytics/requests/dashboard_request_handler.ts:197]: replace with appropriate error message from PM/UX

File: public/components/trace_analytics/requests/request_handler.ts [hunk 4]
```diff
@@ -27,6 +27,24 @@ export function handleDslRequest(http: CoreStart['http'], DSL: any, bodyQuery: a
   if (!bodyQuery.index) {
     body = {...bodyQuery, index: ((mode === 'jaeger' ? JAEGER_INDEX_NAME : DATA_PREPPER_INDEX_NAME)) }
   }
+  if (mode === 'jaeger' && timeout) {
+    const controller = new AbortController();
+    const id = setTimeout(() => controller.abort(), 20000);
```
Comment (joshuali925) [inline public/components/trace_analytics/requests/request_handler.ts:32]: you can show request taking long time toast here (or trigger an callback) and remove abortcontroller. i don't think we should abort the request in 20s, it's limiting the user

File: public/components/trace_analytics/components/dashboard/dashboard_content.tsx [hunk 5]
```diff
@@ -75,59 +84,120 @@ export function DashboardContent(props: DashboardProps) {
...
+          newFilteredService = filter.value;
+          break;
+        }
       }
     }
     setFilteredService(newFilteredService);
-    if (!redirect && indicesExist) refresh(newFilteredService);
-  }, [filters, startTime, endTime, appConfigs]);
+    if (!redirect && ((mode === 'data_prepper' && dataPrepperIndicesExist) || (mode === 'jaeger' && jaegerIndicesExist))) refresh(newFilteredService);
+  }, [filters, startTime, endTime, redirect, mode, dataPrepperIndicesExist, jaegerIndicesExist]);
```
Comment (derek-ho) [inline public/components/trace_analytics/components/dashboard/dashboard_content.tsx:101]: appConfigs is coming from app analytics rendering this component but causes infinite retry with timeout of jaeger query

Comment (derek-ho) [reply public/components/trace_analytics/components/dashboard/dashboard_content.tsx:101]: So not sure if we can remove it

Comment (derek-ho) [reply public/components/trace_analytics/components/dashboard/dashboard_content.tsx:101]: What does appConfigs of App analytics actually contain/will they never change since each app is set and cannot be changed?

Comment (derek-ho) [reply public/components/trace_analytics/components/dashboard/dashboard_content.tsx:109]: Initial testing with dashboard in app analytics seems like it will be ok - this page will be re-rendered on any app config change

File: public/components/trace_analytics/requests/request_handler.ts [hunk 6]
```diff
@@ -19,16 +23,41 @@ export function handleDslRequest(http: CoreStart['http'], DSL: any, bodyQuery: a
...
+  if (!bodyQuery.index) {
+    body = {...bodyQuery, index: ((mode === 'jaeger' ? JAEGER_INDEX_NAME : DATA_PREPPER_INDEX_NAME)) }
+  }
+  if (mode === 'jaeger' && timeout) {
+    const id = setTimeout(() => setShowTimeoutToast!(), 30000);
+    return http
+      .post(TRACE_ANALYTICS_DSL_ROUTE, {
+        body: JSON.stringify(body),
+      }).then((res) => {
+        return res;
```
Comment (joshuali925) [inline public/components/trace_analytics/requests/request_handler.ts:36]: this ".then res return res" is redundent

File: public/components/trace_analytics/requests/request_handler.ts [hunk 7]
```diff
@@ -19,16 +23,41 @@ export function handleDslRequest(http: CoreStart['http'], DSL: any, bodyQuery: a
     if (DSL.query.bool.minimum_should_match)
       bodyQuery.query.bool.minimum_should_match = DSL.query.bool.minimum_should_match;
   }
+  let body = bodyQuery;
+  if (!bodyQuery.index) {
+    body = {...bodyQuery, index: ((mode === 'jaeger' ? JAEGER_INDEX_NAME : DATA_PREPPER_INDEX_NAME)) }
+  }
+  if (mode === 'jaeger' && timeout) {
```
Comment (joshuali925) [inline public/components/trace_analytics/requests/request_handler.ts:30]: remove `mode === 'jaeger'` since you are checking `timeout`? or any reason to check both?

File: public/components/trace_analytics/components/traces/span_detail_table.tsx [hunk 8]
```diff
@@ -10,13 +10,15 @@ import moment from 'moment';
 import React, { useCallback, useEffect, useMemo, useState } from 'react';
```
Comment (pjfitzgibbons) [inline public/components/trace_analytics/components/traces/span_detail_table.tsx:10]: I'll suggest you create a fast-follow Issue/PR for refactor (2.6).

I find it really hard to follow the flow of this component with so many internals switched between Jaeger/Dataprepper with a `mode = .. ? ` tertiary test.

It might be an educational architecture exercise, as well as grealy clarify the code, to refactor this so that either the relevant data points can be gathered and sent as props to <SpanDetailTable>
... OR make <SpanDetailTable> a wrapper and put the relevant detail into a child component.

In the latter case, the high-level jsx might be :

```
<SpanDetailTable>
  {if mode == 'jaeger'}
    <JaegerDetailTable ... >
  { else }
    <DataPrepperDetailTable>
  { end }
</SpanDetailTable>

No sure what is the more-right architecture (opportunity for a spike!), hopefully the end-result will be a far clearer to read module.

File: public/components/trace_analytics/components/dashboard/dashboard_content.tsx (hunk 5 above)
Comment (joshuali925) [reply public/components/trace_analytics/components/dashboard/dashboard_content.tsx:110]: its app filters

https://github.com/opensearch-project/dashboards-observability/blob/3dd96c5f7d18ea27c10745603b74a790dd9ae43e/public/components/trace_analytics/components/common/filters/filters.tsx#L36

Comment (joshuali925) [reply public/components/trace_analytics/components/dashboard/dashboard_content.tsx:110]: what triggered the infinite retry?

File: public/components/trace_analytics/components/traces/span_detail_table.tsx (hunk 8 above)
Comment (derek-ho) [reply public/components/trace_analytics/components/traces/span_detail_table.tsx:10]: Good suggestion - this would make it more extensible if we add more types in the future. Will note it down, thanks!

File: public/components/trace_analytics/requests/request_handler.ts [hunk 9]
```diff
@@ -19,16 +23,41 @@ export function handleDslRequest(http: CoreStart['http'], DSL: any, bodyQuery: a
...
+    const id = setTimeout(() => setShowTimeoutToast!(), 30000);
+    return http
+      .post(TRACE_ANALYTICS_DSL_ROUTE, {
+        body: JSON.stringify(body),
+      }).then((res) => {
+        return res;
+      })
+      .catch((error) => {
+        console.log(error)
```
Comment (ps48) [inline public/components/trace_analytics/requests/request_handler.ts:39]: Can you please change `console.log`to `console.error`?

File: public/components/trace_analytics/requests/queries/services_queries.ts [hunk 10]
```diff
@@ -204,7 +230,117 @@ export const getServiceMetricsQuery = (DSL: any, serviceNames: string[], map: Se
...
+                              },
+                            },
+                          },
+                        },
+                      },
+                      // {
+                      //   terms: {
+                      //     name: targetResource,
+                      //   },
+                      // },
```
Comment (ps48) [inline public/components/trace_analytics/requests/queries/services_queries.ts:267]: Let's remove the comments as well

Comment (derek-ho) [reply public/components/trace_analytics/requests/queries/services_queries.ts]: Fixed in [45cdc1e](https://github.com/opensearch-project/dashboards-observability/pull/150/commits/45cdc1e63436848316f6aa1ab5f6f352fcfa9a11)

File: public/components/trace_analytics/requests/request_handler.ts (hunk 9 above)
Comment (derek-ho) [reply public/components/trace_analytics/requests/request_handler.ts]: Fixed in [45cdc1e](https://github.com/opensearch-project/dashboards-observability/pull/150/commits/45cdc1e63436848316f6aa1ab5f6f352fcfa9a11)

File: public/components/trace_analytics/requests/request_handler.ts (hunk 7 above)
Comment (derek-ho) [reply public/components/trace_analytics/requests/request_handler.ts]: True. Fixed in [45cdc1e](https://github.com/opensearch-project/dashboards-observability/pull/150/commits/45cdc1e63436848316f6aa1ab5f6f352fcfa9a11)

File: public/components/trace_analytics/requests/request_handler.ts (hunk 6 above)
Comment (derek-ho) [reply public/components/trace_analytics/requests/request_handler.ts]: Forgot to take that out when I was playing around with timeout, thanks for catching it 
[45cdc1e](https://github.com/opensearch-project/dashboards-observability/pull/150/commits/45cdc1e63436848316f6aa1ab5f6f352fcfa9a11)

File: public/components/trace_analytics/components/dashboard/dashboard_content.tsx (hunk 5 above)
Comment (joshuali925) [reply public/components/trace_analytics/components/dashboard/dashboard_content.tsx]: adding a comment here for future reference. appConfigs is always an empty array for non-application dashboard, when toast changes and dashboard content re-renders, the array is reallocated and triggers useEffect. setting to a dummy useState array fixes it.

need to (in the future) find out if the empty array is being used, and remove it for non-app dashboard if not used

**Review by joshuali925 — APPROVED** [review]

**Review by ps48 — APPROVED** [review]

File: public/components/trace_analytics/components/traces/span_detail_table.tsx (hunk 8 above)
Comment (ps48) [reply public/components/trace_analytics/components/traces/span_detail_table.tsx:10]: @pjfitzgibbons I was thinking on the same lines. This way it is easier for future additions.

**Comment by vvvprabhakar** [comment]
Pretty Useful feature 
Would love to know to more about what are the other roadmap features for trace data with this plugin

**Comment by derek-ho** [comment]
@vvvprabhakar we are still coming up with a roadmap for this feature set - the main one being service map support with spark jobs, but feel free to create issues for features you would like to see/prioritized, thanks!

File: public/components/trace_analytics/requests/dashboard_request_handler.ts (hunk 3 above)
Comment (dblock) [reply public/components/trace_analytics/requests/dashboard_request_handler.ts:197]: @derek-ho @joshuali925 @ps48 

This almost shipped. This PR should not have been merged without addressing this problem.
